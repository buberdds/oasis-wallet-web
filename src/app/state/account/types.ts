import { Transaction } from 'app/state/transaction/types'
import { ErrorPayload } from 'types/errors'
import { StringifiedBigInt } from 'types/StringifiedBigInt'

export interface BalanceDetails {
  available: StringifiedBigInt | null
  debonding: StringifiedBigInt | null
  delegations: StringifiedBigInt | null
  total: StringifiedBigInt | null
}

export interface Allowance {
  address: string
  amount: StringifiedBigInt
}

export interface Account extends BalanceDetails {
  address: string
  allowances?: Allowance[]
}

/* --- STATE --- */
export interface AccountState extends Account {
  loading: boolean
  accountError?: ErrorPayload
  transactions: Transaction[]
  transactionsError?: ErrorPayload
}
