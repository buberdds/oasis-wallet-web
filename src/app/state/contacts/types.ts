export interface Contact {
  address: string
  name: string
}

/* --- STATE --- */
export interface ContactsState {
  [key: string]: Contact
}
