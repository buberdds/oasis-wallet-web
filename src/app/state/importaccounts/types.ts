import { ErrorPayload } from 'types/errors'
import { WalletBalance, WalletType } from '../wallet/types'

/* --- STATE --- */
export interface ImportAccountsListAccount {
  address: string
  balance: WalletBalance
  path: number[]
  privateKey?: string
  publicKey: string
  selected: boolean
  type: WalletType
}

export enum ImportAccountsStep {
  OpeningUSB = 'opening_usb',
  LoadingAccounts = 'loading_accounts',
  LoadingBalances = 'loading_balances',
  Done = 'done',
}

export interface ImportAccountsState {
  accounts: ImportAccountsListAccount[]
  deviceRequested: boolean
  error?: ErrorPayload
  showAccountsSelectionModal: boolean
  step?: ImportAccountsStep
}
