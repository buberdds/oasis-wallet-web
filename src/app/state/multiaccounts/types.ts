import { ErrorPayload } from 'types/errors'
import { WalletBalance } from '../wallet/types'

/* --- STATE --- */
export interface MultiAccountsListAccount {
  publicKey: string
  address: string
  path: number[]
  balance: WalletBalance
  selected: boolean
}

export enum MultiAccountsStep {
  OpeningUSB = 'opening_usb',
  LoadingAccounts = 'loading_accounts',
  LoadingBalances = 'loading_balances',
  Done = 'done',
}

export interface MultiAccountsState {
  accounts: MultiAccountsListAccount[]
  error?: ErrorPayload
  mnemonic?: string
  step?: MultiAccountsStep
}
