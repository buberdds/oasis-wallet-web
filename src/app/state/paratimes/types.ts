import { StringifiedBigInt } from 'types/StringifiedBigInt'
import { ErrorPayload } from 'types/errors'
import { ParaTime } from '../../../config'

export enum TransactionFormSteps {
  TransferType = 'transferType',
  ParaTimeSelection = 'paraTimeSelection',
  TransactionRecipient = 'transactionRecipient',
  TransactionAmount = 'transactionAmount',
  TransactionConfirmation = 'transactionConfirmation',
  TransactionSummary = 'transactionSummary',
  TransactionError = 'transactionError',
}

export enum TransactionTypes {
  Deposit = 'deposit',
  Withdraw = 'withdraw',
}

export interface TransactionForm {
  amount: string
  confirmation: boolean
  feeAmount: string
  feeGas: string
  paraTime?: ParaTime
  privateKey: string
  recipient: string
  type: TransactionTypes | undefined
}

export interface ParaTimesState {
  balance: StringifiedBigInt
  isLoading: boolean
  transactionError?: ErrorPayload
  transactionForm: TransactionForm
  transactionFormStep: TransactionFormSteps
}

export type Runtime = {
  address: string
  id: string
  decimals: number
}

export type ParaTimeTransaction = Pick<
  TransactionForm,
  'amount' | 'privateKey' | 'feeAmount' | 'feeGas' | 'recipient' | 'type'
>
