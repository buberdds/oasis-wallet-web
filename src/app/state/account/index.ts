import { PayloadAction } from '@reduxjs/toolkit'
import { Transaction } from 'app/state/transaction/types'
import { createSlice } from 'utils/@reduxjs/toolkit'
import { AccountState, Account } from './types'

export const initialState: AccountState = {
  address: '',
  liquid_balance: 0,

  accountError: null,
  transactions: [],
  transactionsError: null,
  loading: true,
}

const slice = createSlice({
  name: 'account',
  initialState,
  reducers: {
    clearAccount(state, action: PayloadAction<void>) {
      Object.assign(state, initialState)
    },
    fetchAccount(state, action: PayloadAction<string>) {},
    accountLoaded(state, action: PayloadAction<Account>) {
      state.accountError = null
      Object.assign(state, action.payload)
    },
    accountError(state, action: PayloadAction<string>) {
      state.accountError = action.payload
    },
    transactionsLoaded(state, action: PayloadAction<Transaction[]>) {
      state.transactionsError = null
      state.transactions = action.payload
    },
    transactionsError(state, action: PayloadAction<string>) {
      state.transactionsError = action.payload
    },
    setLoading(state, action: PayloadAction<boolean>) {
      state.loading = action.payload
    },
  },
})

export const { actions: accountActions } = slice

export default slice.reducer
