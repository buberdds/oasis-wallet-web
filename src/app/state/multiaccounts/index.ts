import { PayloadAction } from '@reduxjs/toolkit'
import { ErrorPayload } from 'types/errors'
import { createSlice } from 'utils/@reduxjs/toolkit'
import { MultiAccountsListAccount, MultiAccountsState, MultiAccountsStep } from './types'

export const initialState: MultiAccountsState = { accounts: [] }

const slice = createSlice({
  name: 'multiAccounts',
  initialState,
  reducers: {
    clear(state, action: PayloadAction<void>) {
      state.accounts = []
      state.error = undefined
      state.mnemonic = undefined
      state.step = undefined
    },
    enumerateAccountsFromLedger(state, action: PayloadAction<void>) {
      state.step = undefined
      state.accounts = []
    },
    enumerateAccountsFromMnemonic(state, action: PayloadAction<string>) {
      state.step = undefined
      state.accounts = []
    },
    toggleAccount(state, action: PayloadAction<number>) {
      const index = action.payload
      state.accounts[index].selected = !state.accounts[index].selected
    },
    accountsListed(state, action: PayloadAction<MultiAccountsListAccount[]>) {
      state.accounts = action.payload
    },
    setMnemonic(state, action: PayloadAction<string>) {
      state.mnemonic = action.payload
    },
    setStep(state, action: PayloadAction<MultiAccountsStep>) {
      state.step = action.payload
    },
    operationFailed(state, action: PayloadAction<ErrorPayload>) {
      state.error = action.payload
      state.mnemonic = undefined
      state.step = undefined
    },
  },
})

export const { actions: multiAccountsActions } = slice

export default slice.reducer
