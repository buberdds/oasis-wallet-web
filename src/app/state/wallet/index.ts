import { PayloadAction } from '@reduxjs/toolkit'
import { createSlice } from 'utils/@reduxjs/toolkit'
import { LedgerAccount } from '../ledger/types'

import { AddWalletPayload, BalanceUpdatePayload, Wallet, WalletBalance, WalletState } from './types'

export const initialState: WalletState = {
  wallets: {},
  isOpen: false,
  selectedWallet: undefined,
}

const slice = createSlice({
  name: 'wallet',
  initialState,
  reducers: {
    openWalletFromMnemonic(state, action: PayloadAction<string>) {},
    openWalletFromPrivateKey(state, action: PayloadAction<string>) {},
    openWalletsFromLedger(state, action: PayloadAction<LedgerAccount[]>) {},
    selectWallet(state, action: PayloadAction<number>) {},
    closeWallet(state, action: PayloadAction<void>) {},
    fetchWallet(state, action: PayloadAction<Wallet>) {},
    updateBalance(state, action: PayloadAction<BalanceUpdatePayload>) {
      Object.assign(state.wallets[action.payload.walletId].balance, action.payload.balance)
    },
    walletSelected(state, action: PayloadAction<number | undefined>) {
      state.selectedWallet = action.payload
    },
    addWallet(state, action: PayloadAction<AddWalletPayload>) {},
    walletOpened(state, action: PayloadAction<Wallet>) {
      const newWallet = action.payload
      state.wallets[newWallet.id] = Object.assign({}, newWallet)
      state.isOpen = true
    },
    walletClosed(state, action: PayloadAction<void>) {
      // Revert to initial state
      Object.assign(state, initialState)
    },
    walletLoaded(state, action: PayloadAction<WalletBalance>) {
      state.isOpen = true
      // state.wallets[state.selectedWallet!].balance = action.payload
    },
  },
})

export const { actions: walletActions } = slice

export default slice.reducer
