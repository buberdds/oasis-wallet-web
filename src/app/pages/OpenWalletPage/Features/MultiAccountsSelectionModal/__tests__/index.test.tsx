import { cleanup, render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { multiAccountsActions } from 'app/state/multiaccounts'
import { MultiAccountsStep } from 'app/state/multiaccounts/types'
import { walletActions } from 'app/state/wallet'
import * as React from 'react'
import { Provider, useDispatch } from 'react-redux'
import { configureAppStore } from 'store/configureStore'
import { ThemeProvider } from 'styles/theme/ThemeProvider'

import { MultiAccountsSelectionModal } from '..'

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}))

const renderComponent = (store: any, abortFunction = () => {}) =>
  render(
    <Provider store={store}>
      <ThemeProvider>
        <MultiAccountsSelectionModal abort={abortFunction} type="ledger" />
      </ThemeProvider>
    </Provider>,
  )

describe('<MultiAccountsSelectionModal  />', () => {
  let store: ReturnType<typeof configureAppStore>

  beforeEach(() => {
    store = configureAppStore()
    jest.mocked(useDispatch).mockImplementation(() => jest.fn())
  })

  afterEach(() => {
    jest.resetModules()
    jest.resetAllMocks()
    cleanup()
    document.body.innerHTML = ''
  })

  it('should match snapshot', () => {
    const component = renderComponent(store)
    expect(component.baseElement).toMatchSnapshot()
    screen.getByTestId('ledger-open-accounts')
  })

  it('should list the accounts when done', () => {
    const component = renderComponent(store)
    store.dispatch(
      multiAccountsActions.accountsListed([
        {
          address: 'oasis1qzyqaxestzlum26e2vdgvkerm6d9qgdp7gh2pxqe',
          balance: { available: '0', validator: { escrow: '0', escrow_debonding: '0' } },
          path: [44, 474, 0],
          publicKey: '00',
          selected: false,
        },
      ]),
    )

    store.dispatch(multiAccountsActions.setStep(MultiAccountsStep.Done))
    expect(component.getByText('oasis1qzyq...7gh2pxqe')).toBeInTheDocument()
  })

  it('should open the selected accounts', () => {
    const dispatchFn = jest.fn()
    jest.mocked(useDispatch).mockImplementation(() => dispatchFn)

    renderComponent(store)
    store.dispatch(
      multiAccountsActions.accountsListed([
        {
          address: 'oasis1qzyqaxestzlum26e2vdgvkerm6d9qgdp7gh2pxqe',
          balance: { available: '0', validator: { escrow: '0', escrow_debonding: '0' } },
          path: [44, 474, 0],
          publicKey: '00',
          selected: false,
        },
        {
          address: 'oasis1qqv25adrld8jjquzxzg769689lgf9jxvwgjs8tha',
          balance: { available: '0', validator: { escrow: '0', escrow_debonding: '0' } },
          path: [44, 474, 1],
          publicKey: '00',
          selected: false,
        },
      ]),
    )

    store.dispatch(multiAccountsActions.setStep(MultiAccountsStep.Done))
    userEvent.click(screen.getByText('oasis1qzyq...7gh2pxqe'))
    expect(dispatchFn).toHaveBeenLastCalledWith({ payload: 0, type: multiAccountsActions.toggleAccount.type })
    store.dispatch(multiAccountsActions.toggleAccount(0))

    userEvent.click(screen.getByTestId('ledger-open-accounts'))
    expect(dispatchFn).toHaveBeenLastCalledWith(
      expect.objectContaining({ type: walletActions.openWalletsFromLedger.type }),
    )
  })
})
