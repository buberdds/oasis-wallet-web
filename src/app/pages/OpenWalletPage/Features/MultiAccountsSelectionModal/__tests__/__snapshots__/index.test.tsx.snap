// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiAccountsSelectionModal  /> should match snapshot 1`] = `
.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 750px;
  padding: 24px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding-top: 48px;
}

.c9 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  width: 12px;
}

.c8 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #0092f6;
  border-radius: 4px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  opacity: 0.3;
  cursor: default;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
  font-weight: bold;
}

.c8:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c8:focus > circle,
.c8:focus > ellipse,
.c8:focus > line,
.c8:focus > path,
.c8:focus > polygon,
.c8:focus > polyline,
.c8:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c8:focus::-moz-focus-inner {
  border: 0;
}

.c8:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c8:focus:not(:focus-visible) > circle,
.c8:focus:not(:focus-visible) > ellipse,
.c8:focus:not(:focus-visible) > line,
.c8:focus:not(:focus-visible) > path,
.c8:focus:not(:focus-visible) > polygon,
.c8:focus:not(:focus-visible) > polyline,
.c8:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c8:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c10 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  border: 2px solid #0092f6;
  border-radius: 4px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  background-color: #0092f6;
  color: #f8f8f8;
  border-radius: 4px;
  opacity: 0.3;
  cursor: default;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
  font-weight: bold;
}

.c10:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c10:focus > circle,
.c10:focus > ellipse,
.c10:focus > line,
.c10:focus > path,
.c10:focus > polygon,
.c10:focus > polyline,
.c10:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c10:focus::-moz-focus-inner {
  border: 0;
}

.c10:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c10:focus:not(:focus-visible) > circle,
.c10:focus:not(:focus-visible) > ellipse,
.c10:focus:not(:focus-visible) > line,
.c10:focus:not(:focus-visible) > path,
.c10:focus:not(:focus-visible) > polygon,
.c10:focus:not(:focus-visible) > polyline,
.c10:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c10:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c6 {
  margin-top: 0px;
  margin-bottom: 24px;
  font-size: 1;
  line-height: normal;
  max-width: 1200px;
  font-weight: 600;
}

.c0 {
  font-family: Rubik;
  font-size: 18px;
  line-height: 24px;
  background-color: #ffffff;
  color: #444444;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  font-family: Rubik;
  font-size: 18px;
  line-height: 24px;
  background-color: #ffffff;
  color: #444444;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  background: transparent;
  position: relative;
  z-index: 20;
  pointer-events: none;
  outline: none;
  position: fixed;
  top: 0px;
  left: 0px;
  right: 0px;
  bottom: 0px;
}

.c2 {
  position: absolute;
  top: 0px;
  left: 0px;
  right: 0px;
  bottom: 0px;
  background-color: rgba(0,0,0,0.5);
  color: #f8f8f8;
  pointer-events: all;
  will-change: transform;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-height: 48px;
  background-color: #FFFFFF;
  color: #444444;
  outline: none;
  pointer-events: all;
  z-index: 20;
  position: absolute;
  max-height: calc(100% - 0px - 0px);
  max-width: calc(100% - 0px - 0px);
  border-radius: 4px;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  -webkit-animation: hWAzxx 0.2s ease-in-out forwards;
  animation: hWAzxx 0.2s ease-in-out forwards;
}

.c4 {
  width: 0;
  height: 0;
  overflow: hidden;
  position: absolute;
}

@media only screen and (max-width:768px) {
  .c5 {
    padding: 12px;
  }
}

@media only screen and (max-width:768px) {
  .c7 {
    padding-top: 24px;
  }
}

@media only screen and (max-width:768px) {
  .c9 {
    width: 6px;
  }
}

@media only screen and (max-width:768px) {
  .c6 {
    margin-top: 0px;
  }
}

@media only screen and (max-width:768px) {
  .c6 {
    margin-bottom: 12px;
  }
}

@media only screen and (max-width:768px) {
  .c1 {
    position: fixed;
    width: 100%;
    height: 100%;
    min-height: 100vh;
  }
}

@media only screen and (max-width:768px) {
  .c2 {
    position: relative;
  }
}

@media only screen and (max-width:768px) {
  .c3 {
    position: relative;
    max-height: none;
    max-width: none;
    border-radius: 0;
    height: 100vh;
    width: 100vw;
  }
}

<body
  style=""
>
  <div>
    <div
      class="c0"
      style="min-height: 100vh;"
    />
  </div>
  <div>
    <div
      aria-hidden="false"
    >
      <div
        class="c1"
        tabindex="-1"
      >
        <div
          class="c2"
        />
        <div
          class="c3"
          data-g-portal-id="0"
          style="overflow-y: auto;"
        >
          <a
            aria-hidden="true"
            class="c4"
            tabindex="-1"
          />
          <div>
            <div
              class="c5"
            >
              <h1
                class="c6"
                size="1"
              >
                openWallet.multiaccounts.selectWallets
              </h1>
              <div
                class="c7"
              >
                <button
                  class="c8"
                  disabled=""
                  type="button"
                >
                  common.cancel
                </button>
                <div
                  class="c9"
                />
                <button
                  class="c10"
                  data-testid="ledger-open-accounts"
                  disabled=""
                  type="button"
                >
                  openWallet.multiaccounts.openWallets
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
